R/calculate_stay.R

```diff
@@ -0,0 +1,43 @@
+#' Calculate total length of stay
+#'
+#' @description Calculate the total length of stay between start_date and end_date.
+#' If the end_date is missing then use the dummy discharge date.
+#'
+#' @param year The financial year in '1920' format
+#' @param start_date The admission/start date variable. e.g. record_keydate1
+#' @param end_date The discharge/end date variable. e.g record_keydate2
+#' @param sc_qtr The latest submitted quarter. e.g. sc_latest_submission
+#'
+#' @return a [tibble][tibble::tibble-package] with additional variable `stay`.
+#' If there is no end date use dummy discharge to calculate the total length of stay.
+#' If there is no end date but sc_qtr is supplied then set this to the end of quarter.
+#' If the quarter end date < start date and sc_qtr is supplied then set this to the end of next quarter.
+#' @export
+#'
+#' @family date functions
+calculate_stay <- function(year, start_date, end_date, sc_qtr = NULL) {
+
+  # Set Quarters
+  qtr_end <- lubridate::add_with_rollback(lubridate::yq(sc_qtr), lubridate::period(6, "months"))
+  next_qtr <- lubridate::add_with_rollback(lubridate::yq(sc_qtr), lubridate::period(9, "months"))
+
+  if (missing(sc_qtr)) {
+    # Do normal stay calculation
+    dummy_discharge <- dplyr::if_else(
+      is.na(end_date),
+      end_fy(year) + lubridate::days(1),
+      end_date
+    )
+
+    lubridate::time_length(lubridate::interval(start_date, dummy_discharge), unit = "days")
+  } else {
```

Comment:
Rearrange the code to only work out the quarters when needed, and to use the first quarter date to work out the second.

```suggestion
  if (missing(sc_qtr)) {
    # Do normal stay calculation
    dummy_discharge <- dplyr::if_else(
      is.na(end_date),
      end_fy(year) + lubridate::days(1),
      end_date
    )

    lubridate::time_length(lubridate::interval(start_date, dummy_discharge), unit = "days")
  } else {
   # Set Quarters
  qtr_end <- lubridate::add_with_rollback(lubridate::yq(sc_qtr), lubridate::period(6, "months"))
  next_qtr <- lubridate::add_with_rollback(qtr_end , lubridate::period(3, "months"))
  
```